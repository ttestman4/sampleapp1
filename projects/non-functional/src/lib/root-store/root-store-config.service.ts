import { InjectionToken } from '@angular/core';
import { RootStoreConfig } from './root-store.models';

/**
 * This is not a real service, but it looks like it from the outside.
 * It's just an InjectionToken used to import the config object,
 * provided from the outside
 */
export const RootStoreConfigService =
    new InjectionToken<RootStoreConfig>('RootStoreConfig');
