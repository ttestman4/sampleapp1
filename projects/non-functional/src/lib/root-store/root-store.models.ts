export interface RootStoreConfig {
    debugStore: boolean;
}

export class StoreConfig {
    static config: RootStoreConfig = {
        debugStore: false,
    };
}
