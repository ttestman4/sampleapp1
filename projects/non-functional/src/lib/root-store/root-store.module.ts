import { CommonModule } from '@angular/common';
import { InjectionToken, ModuleWithProviders, NgModule, Optional, SkipSelf } from '@angular/core';
import { EffectsModule } from '@ngrx/effects';
import { StoreRouterConnectingModule } from '@ngrx/router-store';
import { META_REDUCERS, StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { RootStoreConfig, StoreConfig } from './root-store.models';
import { createMetaReducers, reducers } from './root-store.reducer';
export { RootStoreConfig } from './root-store.models';

/**
 * This is not a real service, but it looks like it from the outside.
 * It's just an InjectionToken used to import the config object,
 * provided from the outside
 */
export const ROOT_STORE_CONFIG =
  new InjectionToken<RootStoreConfig>('ROOT_STORE_CONFIG');

@NgModule({
  declarations: [],
  imports: [
    CommonModule,
    /**
     * StoreModule.forRoot is imported once in the root module, accepting a reducer
     * function or object map of reducer functions. If passed an object of
     * reducers, combineReducers will be run creating your application
     * meta-reducer. This returns all providers for an @ngrx/store
     * based application.
     */
    StoreModule.forRoot(reducers),

    /**
     * @ngrx/router-store keeps router state up-to-date in the store.
     */
    StoreRouterConnectingModule.forRoot(),

    /**
     * Store devtools instrument the store retaining past versions of state
     * and recalculating new states. This enables powerful time-travel
     * debugging.
     *
     * To use the debugger, install the Redux Devtools extension for either
     * Chrome or Firefox
     *
     * See: https://github.com/zalmoxisus/redux-devtools-extension
     */
    StoreDevtoolsModule.instrument({ logOnly: true }),
    /**
     * EffectsModule.forRoot() is imported once in the root module and
     * sets up the effects class to be initialized immediately when the
     * application starts.
     *
     * See: https://github.com/ngrx/platform/blob/master/docs/effects/api.md#forroot
     */
    EffectsModule.forRoot([]),
  ],
  providers: [
    {
      provide: META_REDUCERS,
      deps: [ROOT_STORE_CONFIG],
      useFactory: createMetaReducers
    }
  ]
})
export class RootStoreModule {
  constructor(@Optional() @SkipSelf() parentModule: RootStoreModule) {
    if (parentModule) {
      throw new Error(
        'RootStoreModule is already loaded. Import it in the AppModule only');
    }
  }
  static forRoot(config?: RootStoreConfig): ModuleWithProviders {
    StoreConfig.config = Object.assign({}, StoreConfig.config, config);
    return {
      ngModule: RootStoreModule,
      providers: [
        {
          provide: ROOT_STORE_CONFIG,
          useValue: config ? config : {}
        },
      ]
    };
  }
  static forChild(): ModuleWithProviders {
    return {
      ngModule: RootStoreModule,
      providers: [
      ]
    };
  }
  static forTestReset() {
  }
}
