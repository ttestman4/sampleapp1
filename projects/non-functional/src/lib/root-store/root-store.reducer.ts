import * as fromRouter from '@ngrx/router-store';
import { Action, ActionReducer, ActionReducerMap } from '@ngrx/store';
/**
 * storeFreeze prevents state from being mutated. When mutation occurs, an
 * exception will be thrown. This is useful during development mode to
 * ensure that none of the reducers accidentally mutates the state.
 */
import { storeFreeze } from 'ngrx-store-freeze';
import { RootStoreConfig } from './root-store.models';

/**
 * As mentioned, we treat each reducer like a table in a database. This means
 * our top level state interface is just a map of keys to inner state types.
 */
export interface State {
    name: string;
    router: fromRouter.RouterReducerState;
}

export const initialState = 'Root Store';

export function reducer(state = initialState, action: Action): string {
    switch (action.type) {
        default: {
            return state;
        }
    }
}
/**
 * Our state is composed of a map of action reducer functions.
 * These reducer functions are called with each dispatched action
 * and the current or initial state and return a new immutable state.
 */
export const reducers: ActionReducerMap<State> = {
    name: reducer,
    router: fromRouter.routerReducer,
};

// console.log all actions
export function logger<T, V extends Action = Action>(nextReducer: ActionReducer<T, V>): ActionReducer<T, V> {
    return (state, action): any => {
        const result = nextReducer(state, action);
        console.groupCollapsed(action.type);
        console.log('prev state', state);
        console.log('action', action);
        console.log('next state', result);
        console.groupEnd();

        return result;
    };
}

/**
 * By default, @ngrx/store uses combineReducers with the reducer map to compose
 * the root meta-reducer. To add more meta-reducers, provide an array of meta-reducers
 * that will be composed to form the root meta-reducer.
 */
export function createMetaReducers(config: RootStoreConfig) {
    return config.debugStore === true
        ? [storeFreeze]
        : [];
}
