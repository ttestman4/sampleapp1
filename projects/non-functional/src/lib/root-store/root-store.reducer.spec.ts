import { initialState, reducer } from './root-store.reducer';

describe('Root Reducer', () => {
    it('Unknown action should return the initial state', () => {
        const action = {} as any;

        const result = reducer(initialState, action);

        expect(result).toMatchSnapshot();
    });

    it('Unknown action should return the initial state for undefined initial state input', () => {
        const action = {} as any;

        const result = reducer(undefined, action);

        expect(result).toMatchSnapshot();
    });
});
