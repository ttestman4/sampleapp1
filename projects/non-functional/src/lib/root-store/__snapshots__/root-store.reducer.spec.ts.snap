// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root Reducer Unknown action should return the initial state 1`] = `"Root Store"`;

exports[`Root Reducer Unknown action should return the initial state for undefined initial state input 1`] = `"Root Store"`;
